src/hotspot/cpu/riscv/assembler_riscv_c.hpp

```diff
@@ -0,0 +1,581 @@
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef CPU_RISCV_ASSEMBLER_RISCV_CEXT_HPP
+#define CPU_RISCV_ASSEMBLER_RISCV_CEXT_HPP
+
+private:
+  bool _in_compressible_region;
+public:
+  bool in_compressible_region() const { return _in_compressible_region; }
+  void set_in_compressible_region(bool b) { _in_compressible_region = b; }
+public:
+
+  // C-Ext: If an instruction is compressible, then
+  //   we will implicitly emit a 16-bit compressed instruction instead of the 32-bit
+  //   instruction in Assembler. All below logic follows Chapter -
+  //   "C" Standard Extension for Compressed Instructions, Version 2.0.
+  //   We can get code size reduction and performance improvement with this extension,
+  //   considering the reduction of instruction size and the code density increment.
+
+  // Note:
+  //   1. When UseRVC is enabled, 32-bit instructions under 'CompressibleRegion's will be
+  //      transformed to 16-bit instructions if compressible.
+  //   2. C-Ext's instructions in Assembler always begin with 'c_' prefix, as 'c_li',
+  //      but most of time we have no need to explicitly use these instructions.
+  //   3. In some cases, we need to force using one instruction's uncompressed version,
+  //      for instance code being patched should remain its general and longest version
+  //      to cover all possible cases, or code requiring a fixed length.
+  //      So we introduce 'UncompressibleRegion' to force instructions in its range
+  //      to remain its normal 4-byte version.
+  //      An example:
+  //
+  //        CompressibleRegion cr(_masm);
+  //        __ andr(...);      // this instruction could change to c.and if qualified
+  //        {
+  //          UncompressibleRegion ur(_masm);
+  //          __ andr(...);    // this instruction would remain the normal 32-bit form of andr
+  //        }
+  //
+  //   4. Using -XX:PrintAssemblyOptions=no-aliases could print C-Ext instructions instead of
+  //      normal ones.
+  //
+
+  // C-Ext: extract a 16-bit instruction.
+  static inline uint16_t c_extract(uint16_t val, unsigned msb, unsigned lsb) {
```

Comment:
We definitely need test coverage for those newly added assember functions for this PR.